presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/RuleIndex.java

```diff
@@ -12,28 +12,29 @@
  * limitations under the License.
  */
 
-package com.facebook.presto.matching;
+package com.facebook.presto.sql.planner.iterative;
 
+import com.facebook.presto.matching.Pattern;
 import com.google.common.collect.ImmutableListMultimap;
 import com.google.common.collect.ListMultimap;
 import com.google.common.reflect.TypeToken;
 
 import java.util.Set;
 import java.util.stream.Stream;
 
-public class MatchingEngine<T extends Matchable>
+public class RuleIndex
 {
-    private final ListMultimap<Class, T> matchablesByClass;
+    private final ListMultimap<Class, Rule> rulesByRootType;
```

Comment:
If I get it right, what we'd need is a `ByPatternIndex<T>` with a builder/constructor taking the `getPattern` function for the relevant type. So the usage in `IterativeOptimizer` would be:

```java
    this.ruleIndex = ByPatternIndex.builder(Rule::getPattern)
                .register(newRules)
                .build();
```

and in stats calculator:


```java
    this.ruleIndex = ByPatternIndex.builder(Rule::getPattern)
                .register(newRules)
                .build();
```

Catch me offline to discuss that :)